backend/layers/persistence/persistence_interface.py

```diff
@@ -0,0 +1,223 @@
+from datetime import datetime
+from typing import List, Optional, Iterable
+from backend.layers.common.entities import (
+    CollectionId,
+    CollectionMetadata,
+    CollectionVersion,
+    CollectionVersionId,
+    DatasetArtifact,
+    DatasetArtifactId,
+    DatasetConversionStatus,
+    DatasetId,
+    DatasetMetadata,
+    DatasetProcessingStatus,
+    DatasetStatus,
+    DatasetUploadStatus,
+    DatasetValidationStatus,
+    DatasetVersion,
+    DatasetVersionId,
+)
+
+
+class DatabaseProviderInterface:
+    def create_canonical_collection(self, owner: str, collection_metadata: CollectionMetadata) -> CollectionVersion:
+        """
+        Creates a new canonical collection, generating a canonical collection_id and a new version_id.
+        Returns the newly created CollectionVersion
+        """
+        pass
+
+    def get_collection_version(self, version_id: CollectionVersionId) -> CollectionVersion:
+        """
+        Retrieves a specific collection version by id
+        """
+        pass
+
+    def get_collection_mapped_version(self, collection_id: CollectionId) -> Optional[CollectionVersion]:
+        """
+        Retrieves the latest mapped version for a collection
+        """
+        pass
+
+    def get_all_versions_for_collection(self, collection_id: CollectionId) -> List[CollectionVersion]:
+        """
+        Retrieves all versions for a specific collections, without filtering
+        """
+        pass
+
+    def get_all_collections_versions(self) -> Iterable[CollectionVersion]:
+        """
+        Retrieves all versions of all collections.
+        TODO: for performance reasons, it might be necessary to add a filtering parameter here.
+        """
+        pass
+
+    def get_all_mapped_collection_versions(self) -> Iterable[CollectionVersion]:
+        """
+        Retrieves all the collection versions that are mapped to a canonical collection.
+        """
+        pass
+
+    def delete_canonical_collection(self, collection_id: CollectionId) -> None:
+        """
+        Deletes (tombstones) a canonical collection.
+        """
+        pass
+
+    def save_collection_metadata(
+        self, version_id: CollectionVersionId, collection_metadata: CollectionMetadata
+    ) -> None:
+        """
+        Saves collection metadata for a collection version
+        """
+        pass
+
+    def save_collection_publisher_metadata(self, version_id: CollectionVersionId, publisher_metadata: Optional[dict]) -> None:
+        """
+        Saves publisher metadata for a collection version. Specify None to remove it
+        """
+        pass
+
+    def add_collection_version(self, collection_id: CollectionId) -> CollectionVersion:
+        """
+        Adds a collection version to an existing canonical collection. The new version copies the following data from
+         the previous version: owner, metadata, publisher_metadata, datasets (IDs).
```

Comment:
They do not. I don't think its necessary though, since it's everything except version_id and date-based fields (created_at, published_at) which make intuitive sense not to copy over. Worth updating this comment to be clear about that though rather than listing individual fields, since that is confusing